crates/wasi-common/src/table.rs

```diff
@@ -0,0 +1,111 @@
+use crate::{Error, ErrorExt};
+use std::any::Any;
+use std::cell::{Ref, RefCell, RefMut};
+use std::collections::HashMap;
+
+/// The `Table` type is designed to map u32 handles to resources. The table is now part of the
+/// public interface to a `WasiCtx` - it is reference counted so that it can be shared beyond a
+/// `WasiCtx` with other WASI proposals (e.g. `wasi-crypto` and `wasi-nn`) to manage their
+/// resources. Elements in the `Table` are `Any` typed.
+///
+/// The `Table` type is intended to model how the Interface Types concept of Resources is shaping
+/// up. Right now it is just an approximation.
+pub struct Table {
+    map: HashMap<u32, RefCell<Box<dyn Any>>>,
+    next_key: u32,
+}
+
+impl Table {
+    /// Create an empty table. New insertions will begin at 3, above stdio.
+    pub fn new() -> Self {
+        Table {
+            map: HashMap::new(),
+            next_key: 3, // 0, 1 and 2 are reserved for stdio
+        }
+    }
+
+    /// Insert a resource at a certain index.
+    pub fn insert_at(&mut self, key: u32, a: Box<dyn Any>) {
+        self.map.insert(key, RefCell::new(a));
+    }
+
+    /// Insert a resource at the next available index.
+    pub fn push(&mut self, a: Box<dyn Any>) -> Result<u32, Error> {
+        loop {
+            let key = self.next_key;
+            // XXX this is not correct. The table may still have empty entries, but our
+            // linear search strategy is quite bad
+            self.next_key = self
+                .next_key
+                .checked_add(1)
+                .ok_or_else(|| Error::trap("out of keys in table"))?;
```

Comment:
Incrementing the key is a very good thing; it helps avoid reuse and catch bugs (and we don't have to deal with `select(2)`).

But why not wrap on overflow? If entries are properly closed after use, there wouldn't be anything wrong with that, especially since the map is checked right after (even though it is not necessary before a wrap).

Reply:
Thanks, I fixed this to wrap on overflow, and only trap when the table is full.